### ISSUE TITLE
How to define my own custom meta-Schema (draft-7 version) ?

### ISSUE THREAD

**Issue by mmorel-35** [json-schema-org/json-schema-spec#1007]
Hello,

I'm working an a application where many schema will be submitted so I need to validate them. And I'd like add some extra validations on the initial  draft 7 schema.
For example :
I would like that :
* the "$schema" field to be mandatory.
* every property simple property (not type object) had an examples field 
* every property property had a description field 
* every "Datetime$" property be a { "type" : "string" , "format" : "date-time"}

Can you give me any advice on how to achieve that?
Thank you !

**Comment by jdesrosiers** [comment]
Start by making a copy of the draft-07 schema and give it a unique `$id`.

```json
{
  "$schema": "http://json-schema.org/draft-07/schema",
  "$id": "https://example.com/my-custom-draft-07-schema",
  "definitions": {
    ...original-definitions...
  },
  ...original-schema...
}
```

Since you want different constraints for schemas in different places, you'll need to refactor the schema to a definition so it can be referenced and extended for different situations. Don't forget to change all the recursive references (`{ "$ref": "#" }`) to point to the definition you created (`{ "$ref": "#/definitions/schema" }`).

```json
{
  "$schema": "http://json-schema.org/draft-07/schema",
  "$id": "https://example.com/my-custom-draft-07-schema",
  "alllOf": [{ "$ref": "#/definitions/schema" }],
  "definitions": {
    ...original-definitions-with-modified-$refs...
   "schema": {
      ...original-schema-with-modified-$refs...
    }
  }
}
```

Next you'll need to add another definition specific to property schemas and change the `$ref`s for property schemas to use the definition that was created (`{ "$ref": "#/definitions/property-schema" }`). Don't forget to change `patternProperties` and `additionalProperties` as well as `properties`. Other keywords, such as `anyOf` should continue to reference the generic schema (`{ "$ref": "#/definitions/schema" }`).

```json
{
  "$schema": "http://json-schema.org/draft-07/schema",
  "$id": "https://example.com/my-custom-draft-07-schema",
  "alllOf": [{ "$ref": "#/definitions/schema" }],
  "definitions": {
    ...original-definitions-with-modified-$refs...
    "schema": {
      ...original-schema-with-modified-$refs...
    },
    "property-schema": {
      "allOf": [{ "$ref": "#/definitions/schema" }]
    }
  }
}
```

At this point, the schema has just been refactored. None of the behavior has changed except that you now have a place to put your custom constraints so they only will only apply where you want them to. It should be relatively straightforward to add your constraints from here.


```json
{
  "$schema": "http://json-schema.org/draft-07/schema",
  "$id": "https://example.com/my-custom-draft-07-schema",
  "alllOf": [{ "$ref": "#/definitions/schema" }],
  ...add-root-schema-constraints-here...
  "definitions": {
    ...original-definitions-with-modified-$refs...
    "schema": {
      ...original-schema-with-modified-$refs...
      ...add-global-constraints-here...
    },
    "property-schema": {
      "allOf": [{ "$ref": "#/definitions/schema" }]
      ...add-property-schema-constraints-here...
    }
  }
}
```

**Comment by mmorel-35** [comment]
Thank you @jdesrosiers  for your time and the precision of your answer it's very helpful !

**Comment by lmichelbacher** [comment]
Hi,

I have similar requirements. I want to...

1. ...validate additional constrains on certain properties
2. ...turn off certain keywords, e.g. `if`/`then`/`else`

It seems that the above approach should meet my needs.

I'm struggling to make this work though. I stumbled over a few things:

1. `original-schema-with-modified-$refs`:  does this include the `properties` level or do I add what's under `properties` directly?

2. Do I remove the original `properties` from the top level?

3. > change the `$ref`s for property schemas to use the definition that was created (`{ "$ref": "#/definitions/property-schema" }`)

I don't understand where I'd have to insert `{ "$ref": "#/definitions/property-schema" }`.

4. How do I turn off certain keywords? I've set `"if": false` under `definitions/schema/properties/if` which has no effect.
